src/hypothesis/internal/conjecture/minimizer.py

```diff
@@ -78,55 +86,95 @@ def _shrink_index(self, i, c):
             hbytes([255] * (self.size - i - 1))
         )
 
+    def ddlower(self, c):
+        self.ddfixate(lambda b: min(b, c))
+
+    def ddshift(self):
+        self.ddfixate(lambda b: b >> 1)
+
+    def ddsub(self):
+        self.ddfixate(lambda b: max(0, b - 1))
+
+    def ddfixate(self, f):
+        prev = -1
+        while prev != self.changes:
+            prev = self.changes
+            k = len(self.current)
+            while k > 0:
+                i = 0
+                while i + k <= len(self.current):
+                    self.incorporate(
+                        self.current[:i] +
+                        hbytes(f(b) for b in self.current[i:i + k]) +
+                        self.current[i + k:]
+                    )
+                    i += k
+                k //= 2
```

Comment:
It's based on an algorithm called delta debugging. The idea is that we're trying to fixate where no improvement of a single value is permitted, but trying each value in turn is potentially slow if a lot of them are likely to be successful, so we try doing bulk replacement of values in large blocks first and then gradually shrink the size of the block down to 1.

Normally delta debugging is used with deleting, but it works just as well with rewriting.